libraries/chain/htlc_evaluator.cpp

```diff
@@ -0,0 +1,167 @@
+         // make sure the preimage length is reasonable
+         FC_ASSERT( o.key_size < htlc_options->max_preimage_size, "HTLC preimage length exceeds allowed length" ); 
+         // make sure we have a hash algorithm set
+         FC_ASSERT( o.hash_type != graphene::chain::hash_algorithm::unknown, "HTLC Hash Algorithm must be set" );
+         // make sure sender has the funds for the HTLC
+         FC_ASSERT( db().get_balance( o.source, o.amount.asset_id ) >= (o.amount ), "Insufficient funds" );
+         // make sure sender has the funds for the fee
+         FC_ASSERT( db().get_balance( o.source, asset().asset_id) > o.fee, "Insufficient funds" );
+         return void_result();
+      }
+
+      object_id_type htlc_create_evaluator::do_apply(const htlc_create_operation& o)
+      {
+          try {
+             //FC_ASSERT( db().head_block_time() > HARDFORK_ESCROW_TIME,
+             //           "Operation not allowed before HARDFORK_ESCROW_TIME."); // remove after HARDFORK_ESCROW_TIME
+
+             db().adjust_balance( o.source, -o.amount );
+
+             const htlc_object& esc = db().create<htlc_object>([&]( htlc_object& esc ) {
+                esc.from                  = o.source;
+                esc.to                    = o.destination;
+                esc.amount                = o.amount;
+                esc.preimage_hash		   = o.key_hash;
+                esc.preimage_size		   = o.key_size;
+                esc.expiration		      = db().head_block_time() + o.seconds_in_force;
+                esc.preimage_hash_algorithm = o.hash_type;
+             });
+             return  esc.id;
+
+          } FC_CAPTURE_AND_RETHROW( (o) )
+      }
+
+      template<typename T>
+      bool htlc_redeem_evaluator::test_hash(const std::vector<unsigned char>& incoming_preimage, const std::vector<unsigned char>& valid_hash)
+      {
+         // convert incoming_preimage to an array
+         unsigned char incoming_array[incoming_preimage.size()];
+         for(unsigned i = 0; i < incoming_preimage.size(); ++i)
+            incoming_array[i] = incoming_preimage[i];
+         T attempted_hash = T::hash( (char*)incoming_array, incoming_preimage.size());
+         if (attempted_hash.data_size() != valid_hash.size())
+            return false;
+         char* data = attempted_hash.data();
+         for(size_t i = 0; i < attempted_hash.data_size(); ++i)
+         {
+            if ( ((unsigned char)data[i]) != valid_hash[i])
+               return false;
+         }
+         return true;
+      }
+
+      void_result htlc_redeem_evaluator::do_evaluate(const htlc_redeem_operation& o)
+      {
+    	   htlc_obj = &db().get<htlc_object>(o.htlc_id);
+
+  		   FC_ASSERT(o.preimage.size() == htlc_obj->preimage_size, "Preimage size mismatch.");
+    		FC_ASSERT(db().head_block_time().sec_since_epoch() < htlc_obj->expiration.sec_since_epoch(), "Preimage provided after escrow expiration.");
```

Comment:
That depends on when head_block_time is updated. Is it after all transactions in the block have been processed? If so, I agree that this would be redundant. If you don't know off the top of your head, I'll research.

Reply:
It's updated in `update_global_dynamic_data`, i. e. after tx processing but before `clear_expired_*`.

Reply:
Removed redundant check